### ISSUE TITLE
Build a Docker image: Ansible playbook vs Dockerfile.

### ISSUE THREAD

**Issue by yakutovicha** [aiidalab/aiidalab-docker-stack#243]
Currently, we build the [aiidalab-docker-stack](https://github.com/aiidalab/aiidalab-docker-stack) image on top of the [aiida-core](https://github.com/aiidateam/aiida-core) image. The latter is built on top of [aiida-prerequisites](https://github.com/aiidateam/aiida-prerequisites).

With this approach, it is super hard to quickly change something in the `aiidalab-docker-stack` image, especially when it requires changes in the `aiida-core` or in the `aiida-prerequisites`. Furthermore, we might want to build `aiidalab-docker-stack` on top of the officially provided [jupyter images](https://jupyter-docker-stacks.readthedocs.io/en/latest/using/selecting.html). But those might not be a good base for the `aiida-core`.

Building a Docker container is no different from installing AiiDAlab on a virtual machine or on bare metal. So Ansible appears to me the best way to unify things.

Hereby, I propose to **_build the AiiDAlab docker stack image using Ansible playbook_**.

## Advantages
- We don't have a strict hierarchy of the containers, no need to wait for all the hierarchy to be updated to achieve something.
- While we can adapt the ansible-playbook for a very specific needs of a container, the ansible roles themselves can remain standard. Here is the list of the ansible roles we can reuse:
  - RabbitMQ ([link](https://github.com/geerlingguy/ansible-role-rabbitmq))
  - PostgreSQL ([link](https://github.com/geerlingguy/ansible-role-postgresql))
  - AiiDA-core ([link](https://github.com/marvel-nccr/ansible-role-aiida))
  - AiiDAlab ([link](https://github.com/marvel-nccr/ansible-role-aiidalab)).
- It is a lot easier to change the operating system inside the Docker container (or to support multiple), as the playbook will essentially remain the same.
- We can also use the same playbook to build AiiDAlab on virtual machines or on bare metal.
- While building the container, we will be routinely testing the Ansible roles to install AiiDA/AiiDAlab.
- Quite a significant chunk of work (installing PostgeSQL, RabbitMQ) can be delegated to the external developers who maintain the corresponding roles.

## Open questions.

**Q1**: The new image is built as a single docker layer - which is not ideal. How can this be fixed?

In principle, we can do it with ansible itself. It is possible to `docker commit` a new image after any modification. However, I personally don't have issues with the number of layers. I would rather focus on decreasing its size.

**Q2**: How standard is this approach?
I don't have a good overview here. Just a few links:

Nana Janashia who is also a [docker captain](https://www.docker.com/captains/nana-janashia) talked about the advantages of this approach in [this video](https://www.youtube.com/watch?v=1id6ERvfozo&t=884s) on her channel.

[This]( https://www.youtube.com/watch?v=kkazBPHc4bk) and [this](https://www.youtube.com/watch?v=-ZAgwl3VyuQ) videos show the actual implementation.

Overall my impression is that this approach is no worse than the standard Docker file approach. To me, the benefits we are getting outweigh the possible drawbacks.

## Previous approaches.
Prior to the current approach (`aiida-prerequisites` -> `aiida-core` -> `aiidalab-docker-stack`) we used a different hierarchy:
1. `aiida-docker-base` (Phusion base image + AiiDA)

The idea behind was to be able to use separate images for aiida, PostgreSQL and RabbitMQ and let them interact through Docker Compose. At the same time, the image alone remained unfunctional.

2. `aiida-docker-stack` (based on aiida-docker-base image, but adds [PostgreSQL](https://www.postgresql.org/) and [RabbitMQ](https://www.rabbitmq.com/) servers on top.)

In the end, we decided to revert the order which led us to the current approach.

**Comment by csadorf** [comment]
@yakutovicha Can you please link some references for this approach?

**Comment by yakutovicha** [comment]
> @yakutovicha Can you please link some references for this approach?

yes, the update is in progress.

**Comment by csadorf** [comment]
Realizing that my previous post might give the wrong impression of my overall assessment, I think we should strongly consider this approach.

**Comment by ltalirz** [comment]
I haven't read through everything, just wanted add a couple of thoughts.

Using ansible for building containers is an old idea, the most obvious issue being the lack of caching for intermediate layers you mention [1]. RedHat started with https://github.com/ansible/ansible-container to address this but then decided to discontinue it.
There is the https://github.com/ansible-community/ansible-bender follow-up project, which allows caching at the ansible task level, but also that doesn't seem to be super active through the last years.
For some reason, this approach has not taken off so far, which makes me think there likely are reasons for this (although I can't say I know them).

Some of the advantages you mention also have solutions in the docker world (e.g. using microservices to offload development to maintainers of rabbitmq/postgres/... container images) that do not translate well to our case since JupyterHub currently forces us to bundle all of them into a single container.

In any case, we currently use the ansible role for the Quantum Mobile image to build a corresponding docker image, so it's certainly doable. However, so far, the docker image of the QM has been more of an afterthought rather than the primary output. This is clearly different here.

Some points to think about:

 * You are introducing one additional layer of complexity (ansible). This can make things less stable, more difficult to edit & to understand. Whatever solution you choose, make sure that the developer experience of updating the container image remains at least as good as the current one
 * Before going down this road, be very clear about which is the main problem you want to solve. If it's just about you being blocked in updating the AiiDAlab image hierarchy, there is a very simple solution: release `aiida-prerequisites` and `aiida-core images` with custom tags that you control. No new tooling, just some new branches - that's it.



[1] You seem to be ok with this, but I'm not so sure. Just imagine having to rebuild the entire aiidalab container because you made a mistake in the last task (multiple times)... it would drive me crazy.

**Comment by csadorf** [comment]
Here are some more resources and one potential alternative approache that I found:

- Alternative approach: [Building docker images with ansible and packer](https://getbetterdevops.io/build-docker-images-using-ansible-and-packer/)
- Does not directly touch on the question at hand (maybe indirectly in point 1), but might be good to review regardles: [Docker Anti Patterns](https://codefresh.io/containers/docker-anti-patterns/)
- Also not _directly_ applicable, but [here is argument on SO](https://stackoverflow.com/questions/26110828/should-i-use-dockerfiles-or-image-commits/26111099#26111099) why one would _usually_ avoid `docker commit`.
- [Reddit](https://www.reddit.com/r/docker/comments/jk2rkb/comment/gah3pka/?utm_source=share&utm_medium=web2x&context=3
) [does](https://www.reddit.com/r/docker/comments/jk2rkb/comment/gah9rhf/?utm_source=share&utm_medium=web2x&context=3
) [not](https://www.reddit.com/r/docker/comments/j6nf82/comment/g7zxpst/?utm_source=share&utm_medium=web2x&context=3
) like this approach and provides some arguments as to why this might be a bad idea. I would caution against overvaluing this. Social media overemphasises loud, negative opinions and I am not sure that the arguments brought forth fully apply in this case, but we should either way consider them and also be able to defend this approach against them.

I fully echo @ltalirz 's comment. Especially this point:

> This approach has not taken off so far, which makes me think there likely are reasons for this (although I can't say I know them).

There is very little evidence for this approach to be used by a significant number of users. That does not mean it is a bad approach _per se_, but it is certainly a red flag and – like was pointed out previously – generates additional documentation burden and potential complexity.

**Comment by yakutovicha** [comment]
Thanks for your comments, guys! Here are some follow-ups.

>  I am going to discuss this approach strictly with respect to the question of using ansible to build the image as opposed to using a standard Dockerfile.

Indeed, the discussion is strictly about this: building an image using Ansible vs a standard docker file. I just changed the title of the issue to better reflect the purpose of the discussion.

> > * We don't have a strict hierarchy of the containers, no need to wait for all the hierarchy to be updated to achieve something.
> 
> I don't fully understand the argument. The question at hand is whether we use a Dockerfile or ansible to build the image. In both cases, we need a base image and in both cases, we are free to use a different base image to avoid this problem. This is in my opinion not really an advantage of using this approach.

Well, this is actually related, and here is how. We cannot _just_ pickup a base image and be happy with it. Any base image will have certain tools installed (PostgreSQL base image will have PostgreSQL, Jupyter base image will have Jupyter, AiiDA base image has AiiDA) and others not. So the AiiDAlab docker file will have to be adapted quite a bit in case we decide to change the base image. On top of that, our current approach with `prerequisites` -> `aiida-core` -> `aiidalab` won't let us simply replace the base image. For example, none of the Jupyter images can be served as the base for `prerequisites`/`aiida-core` images.

The only solution I can see here is to move everything into a single Docker file and decouple the `aiidalab` image from `aiida-core`. But in this case, we can forget about any reuse of code.

At the same time, using Ansible would let us re-use the roles keeping the playbooks independent.

## P.S.

Of course, the ideal solution here is Docker compose where every image has its own responsibility - that is much more in line with what Docker is supposed to be. But, Jupyterhub can't deal with Docker compose. 

So we are currently kind of obliged to use Docker container as VM despite the drawbacks it brings.

> I don't think we should conflate (Docker) containers and VMs, they have different purposes and the fact that we are using containers like VMs is actually a reason I would like to move away from that model if possible.

I agree, but for this, we should either wait for Jupyterhub to support docker-compose or move away from using Jupyter as our base technology.

**Comment by yakutovicha** [comment]
> > * It is a lot easier to change the operating system inside the Docker container (or to support multiple), as the playbook will essentially remain the same.
> 
> Not sure what you mean by "changing" here.

Changing means using another base image built with another operating system.

>  If we are talking about the build process, we would still just change the base image, so not sure why this approach would have an advantage.

Different operating systems would require adaptation of the installation/configuration scripts. Very often ansible roles are able to deal with that.


> Sorry, misread that earlier. Sure, but I would argue that the roles should be tested regardless of how we build this particular image.

I agree with you. Still, I believe the actual use of the roles will make them more versatile/robust.


> We could just use the officially maintained docker images for those services instead. I don't see the clear advantage.

It is not clear to me how can we use them? If you refer to the docker-compose approach - it is not possible with JupyterHub. Unless we move away from JupyterHub - but this is not happening soon.

**Comment by csadorf** [comment]
> Well, this is actually related, and here is how. We cannot _just_ pickup a base image and be happy with it. Any base image will have certain tools installed (PostgreSQL base image will have PostgreSQL, Jupyter base image will have Jupyter, AiiDA base image has AiiDA) and others not.

I am fully aware that a change of the base image requires some adaptations – that's clear. I just don't think that a switch to ansible would make that _significantly_ easier. Sure, some individual steps might get automatically skipped, e.g., if Jupyter is already installed, but in practice, the scripts will still have to be adjusted for a change of the base image.

Like @ltalirz pointed out, you can always create a branch for the base image if we really can't work with the main releases.

> It is not clear to me how can we use them? If you refer to the docker-compose approach - it is not possible with JupyterHub. Unless we move away from JupyterHub - but this is not happening soon.

As you know, JH has some functions to host services etc. and I would be surprised if there were no developments at all to address our use case. I would rather invest some of our time in exploring those than keep pushing the current monolithic model.

**Comment by yakutovicha** [comment]
@csadorf, I will keep replying to the other comments that I haven't replied yet. Once this is done I will come back to your latest comment.

**Comment by yakutovicha** [comment]
> I have not yet formed a conclusive opinion on this approach, but I would like to point out that there are indeed some significant drawbacks:
> 
> 1. As mentioned before, using `docker commit` is a fairly non-standard approach to building docker images. The [official Docker docs](https://docs.docker.com/engine/reference/commandline/commit/#extended-description) recommend to use Dockerfiles over `docker commit`.

Yes, it is non standard. 

> 2. It increases complexity for image building: instead of a simple `docker build .`, developers must setup a complete ansible tool chain.

I would say that the added complexity is to install ansible and the roles:

Ansible approach:
```
pip install ansible
ansible-galaxy role install -r requirements.yml # to install the roles
ansible-playbook docker-image-playbook.yaml 
```

Docker container approach:
```
docker build . -t aiidalab/aiidalab-docker-stack
```

IMO, that is not a big difference. So I wouldn't classify it as a significant drawback.


> 3. It is harder to reproduce: The roles are slightly less transparent compared to showing the actual commands that are used to build the image.

I don't agree, actually. When building an ansible playbook, one can be as explicit and as transparent as one wants. Organising tasks into roles is a feature, not a requirement.

> 4. Fewer docker layers makes caching harder which will increase CI build times and download sizes for users.

That is a drawback for me as well. How significant is this - I don't know.

> 5. A non-standard approach comes with increased documentation burden for maintainers.

Here, I can't really comment. I would need to try it to understand how much additional burden it will be. I am willing to try this approach, but I also don't want to waste my time if it becomes clear that the ansible approach is no go.

**Comment by yakutovicha** [comment]
> [1] You seem to be ok with this, but I'm not so sure this works in practice. Just imagine having to rebuild the entire aiidalab image because you made a mistake in the last task (multiple times)... it would drive me crazy.

When preparing #242 I was building the container. I obviously did run into such issues. I think it is quite doable, but the approach is different. In the case of Ansible the container that is being modified remains running. I was entering that container and was trying to check why certain (typically the last) step didn't work.

On the other hand, I clearly remember the issues with building a Docker image: you run into a problem and you can't immideately check what is going on. The way to have a look was to stop building the image at a certain point (by commenting the rest of the Dockerfile), start the just-produced image and enter the running container.

So I wouldn't be so sure about that point.


> Some of the advantages you mention also have solutions in the docker world (e.g. using microservices to offload development to maintainers of rabbitmq/postgres/... container images) that unfortunately do not translate well to our case since JupyterHub currently forces us to bundle all of them into a single container.

yes


> * You are introducing one additional layer of complexity (ansible). This can make things less stable, more difficult to edit & to understand. Whatever solution you choose, make sure that the developer experience of updating the container image remains at least as good as the current one

I am fully with you here. The additional complexity appears to me not that significant, but I can underestimate it. What I can ensure you that, in case we choose this approach, I will make sure to have everything clearly documented in the readme. In any case I am not merging the #242 without an approval :).

**Comment by yakutovicha** [comment]
> @yakutovicha Before, we make a decision here, is this effort embedded in the larger objective of making AiiDAlab compatible with JupyterLab or is this completely unrelated?

At least in my mind, this is an important prerequisite. I think we need to clearly understand how to build an image before moving forward.

> Because this is what we had agreed upon to focus on next.

I would like to emphasize one thing: I don't think we have agreed on it. My understanding was that we are going to access the feasibility of doing the transition. If changing to JupyterLab would require drastic changes of the interface - this will have to be discussed with a broader audience.

Jupyter Notebook itself is going to be built on JupyterLab components (see the [EAP](https://jupyter.org/enhancement-proposals/79-notebook-v7/notebook-v7.html)) and this seems to be like a viable option to me that should at least be discussed.

**Comment by csadorf** [comment]
> > @yakutovicha Before, we make a decision here, is this effort embedded in the larger objective of making AiiDAlab compatible with JupyterLab or is this completely unrelated?
> 
> At least in my mind, this is an important prerequisite. I think we need to clearly understand how to build an image before moving forward.

Ok, thanks for the clarification.

> > Because this is what we had agreed upon to focus on next.
> 
> I would like to emphasize one thing: I don't think we have agreed on it. My understanding was that we are going to access the feasibility of doing the transition. If changing to JupyterLab would require drastic changes of the interface - this will have to be discussed with a broader audience.
> 
> Jupyter Notebook itself is going to be built on JupyterLab components (see the [EAP](https://jupyter.org/enhancement-proposals/79-notebook-v7/notebook-v7.html)) and this seems to be like a viable option to me that should at least be discussed.

I think we need to have a separate discussion to clarify what exactly I mean by "compatibility with JupyterLab". 🙂

**Comment by yakutovicha** [comment]
> I think we need to have a separate discussion to clarify what exactly I mean by "compatibility with JupyterLab". 🙂

For the record: we are going to have a separate discussion about it. It would probably make sense to move this question to a separate issue and continue the discussion there.

**Comment by csadorf** [comment]
@yakutovicha Is it fair to say that this discussion and the related PR can be closed for now?

**Comment by yakutovicha** [comment]
I think so, yes
